### PR TITLE
MAINT:linalg: Expose Cython functions for generic use

### LINKED ISSUE #16143: ENH: Expose Cythonized functions to other pyx files through cimport
This is more of a learning issue that I have been trying out for a few weeks now but couldn't have any progress. What I would like to have is 
 - expose a Cython API for linalg that is similar to `cython_blas`, `cython_lapack` or `cython_special`. 
 - or at least make certain Cythonized functions common to all scipy functions. 

I think I can say that I really don't understand the machinery `_generate_pyx.py` or the `custom_target` of meson yet. 

#### A particular case study: 

In `matfuncs_expm.pyx.in` file there is a function that is a fast transpose for memoryviews. What I wanted to have is to make it available through `_cythonized_linalg_utils.pyx` and `cimport` it via other pyx files. To do that it turns out that I need a pxd file for the definitions since I am constantly getting 

```
from ._cythonized_array_utils cimport swap_c_and_f_layout
^
------------------------------------------------------------

scipy\linalg\_matfuncs_expm.pyx:31:0: 'scipy\linalg\_cythonized_array_utils.pxd' not found
```

But processing pxd file is quite the black magic both in distutils and also meson. Or put better, I really don't understand it at all. And I would like to throw the towel after wasting a lot of time. 

The resulting structure I tried to have is 

```
-- __cythonized_array_utils  #  this namespace holds the cimportable 'swap_c_f_layout' function
 | ...
 | ...
 |- _matfuncs_expm.pyx.in #  this file cimports the 'swap_c_f_layout'
 |- ...
```

The optimal outcome is to expose a Cython `cython_linalg_utils` API however that turned out to be even more complicated hence the case study above is already enough for me. Maybe it can be also beneficial for others in the future to expose a Cython API in other modules.

#### What I have tried

There are many permutations I've tried but the generic attempt was to create a pxd file and change `setup.py` to

```python
    config.add_data_files('_cythonized_array_utils.pxd')
    config.add_extension('_cythonized_array_utils',
                         sources=['_cythonized_array_utils.c',
                                        '_cythonized_array_utils.pxd'])  #  also tried the "depends" etc. but
                                                                         #  documentation is awful for distutils on these
```

### PR DESCRIPTION
**Pull request by ilayn**
Closes #16143 

Certain low-level Cython functions are useful for multiple functions, such as checking for triangular-ness, or symmetry/Hermitian structure. 

With generous help of @ev-br hopefully this PR manages the cimportable cdef'd functions. 

@rgommers I think this needs you attention since we butchered our way out for the Meson stuff.

### CODE REVIEW

File: scipy/linalg/meson.build [hunk 1]
```diff
@@ -222,6 +222,30 @@ cython_lapack = py3.extension_module('cython_lapack',
...
 
+
+py3.install_sources(
+  ['_cythonized_array_utils.pyi', '_cythonized_array_utils.pxd'],
+  pure: false,
+  subdir: 'scipy/linalg',
+)
+
+# This is needed to add pxd files into build dir ala "cp"
+# See
```
Comment (ilayn) [inline scipy/linalg/meson.build:233]: bah I forgot to add the actual link to `copy : true` warning creation

File: scipy/linalg/setup.py [hunk 2]
```diff
@@ -146,8 +146,14 @@ def configuration(parent_package='', top_path=None):
...
                          sources=['_decomp_update.c'])
 
+    config.add_data_files('_cythonized_array_utils.pxd')
+    
     config.add_extension('_cythonized_array_utils',
-                         sources=['_cythonized_array_utils.c'])
+                         sources=['_cythonized_array_utils.c'],
+                         depends=['_cythonized_array_utils.pyx',
+                                  '_cythonized_array_utils.pxd'],
+                                  include_dirs=['.']
```
Comment (ilayn) [inline scipy/linalg/setup.py:155]: I am not sure if this line is needed. I got varying results and I certainly don't understand anything that is going on here. But this started to give consistent results.

Comment (rgommers) [reply scipy/linalg/setup.py:155]: Yes indeed, it's needed with `distutils`. For Meson it never is; the current source directory is automatically added.

**Comment by rgommers** [comment]
> @rgommers I think this needs you attention since we butchered our way out for the Meson stuff.

No worries, I pushed a fix. Cython + generated sources is still a bit fiddly; the diff in this PR now is hopefully clear.

**Comment by ilayn** [comment]
This is apart from the irrelevant test ready to merge.

**Comment by ilayn** [comment]
Since it doesn't have new functionality and relevant residents checked them out, I'll click the button. We still have time to rearrange things in case needed. Thanks @ev-br @rgommers !

**Pull request merged**